### PR TITLE
Add "managed docker compose" deploy type for managed instances

### PR DESCRIPTION
**Pull request by courier-new**
Part 1 for https://github.com/sourcegraph/sourcegraph/issues/36920.

The Batch Changes team is interested in configuring a setting to be default `false` for managed instances but default `true` for everyone else.

This PR adds a new value for the `DEPLOY_TYPE` environment variable to indicate an instance is managed by Sourcegraph. Since today, managed instances are all run with docker-compose, the value is named as such: "managed-docker-compose." This gives us the flexibility to support additional environment types for managed instances in the future. A corresponding method, `IsManagedInstance`, is added to enable that all future types of managed instances can also be treated the same by the backend code when it makes sense to do so, such as for our team's use case.

All other changes made here are meant to mirror those from https://github.com/sourcegraph/sourcegraph/pull/32407.

## Test plan

I'm not very familiar with this part of the codebase and would appreciate recommendations for the best way to test this. 🙂

### CODE REVIEW

**Review by michaellzc — CHANGES_REQUESTED** [review]
The PR looks good. However,

I think this is better solved by using [feature flags](https://docs.sourcegraph.com/dev/how-to/use_feature_flags) instead of abusing the `DEPLOYMENT_TYPE` env var.

cloud-devops has a CLI to easily mutate an instance configuration. (https://github.com/sourcegraph/deploy-sourcegraph-managed/blob/main/util/cmd/mg_sync_alerts.go)

If batches can tell us what flag to enable (even better, show us the sample `updateFeatureFlag` mutation query!), we can enable the flag on all instances for you.


Also, `IsManagedInstance` sounds really dangerous to me, I would prefer the flag to be description of the behaviour, i.e. `hide_run_batch_spec_button` instead of `is_managed_instance`.

Side note for Cloud V2, I can imagine we would like to utilize feature flags more often to tune the behaviour for individual customers. So it's likely the new Cloud team will need to build something to manage all these flags at scale

**Comment by michaellzc** [comment]
cc @bobheadxi you implemented the feature flag system so figure you may want to chime in as well

**Comment by eseliger** [comment]
> I think this is better solved by using [feature flags](https://docs.sourcegraph.com/dev/how-to/use_feature_flags) instead of abusing the DEPLOYMENT_TYPE env var.

I agree it feels a little like abuse of that variable. Looking at the list `helm` also doesn't fit in perfectly but probably still slightly better than this.
I think that's reasonable and we should probably make this be a second variable, that may or may not live in this package, but is fed from a different data source.

Our need for this is a bit business-driven, until we have sorted out billing and supporting maintenance of executors on MIs, we are not able to sell this product that we want to feature-gate to MI customers. It is critical that this gate is always true, we don't want to have customers reach out and be discouraged or them derailing sales conversations because they can't access a certain feature that is showing in their UI.

If we just set an env var that we just have to put into the template for managed instances once and get consistently configured instances for _every_ customer that seems like a good and cheap way to achieve what our needs are here, it also seems fairly simple to implement (but please prove me wrong :D). To be clear, this will eventually _toggle off a feature_, but it is not implemented using _feature flags_ which you linked (this can be confusing!). Would that change your mind? 🤔

**Comment by courier-new** [comment]
Thanks for the quick review, Michael! (And thank you Erik for adding some color to this!!)

The other motivation for this solution was actually what you touched on with your concern about `IsManagedInstance`. We thought that it would actually be _more_ useful to have something more generally-applicable than just a Batch Changes-specific setting. It sounds like not having a way to detect a managed instance has been annoying for teams before, and we do already employ a similar treatment for dotcom with `SourcegraphDotComMode`.

Anyway, just another thought for consideration. I'm all for doing this the right way, I'm just not sure the right way is a feature flag. 🙂

**Comment by daxmc99** [comment]
> we do already employ a similar treatment for dotcom with `SourcegraphDotComMode`

@rafax has mentioned before this is slightly an anti-pattern since it results in code that starts to diverge depending on "if" it was intended to be cloud-specific. With cloud no longer being a primary focus, some of this behavior will need to be re-developed leveraging feature flags so we can precisely toggle it on/off.

I also think the specific behavior we are aiming for here should would be better accomplished via feature flags as @michaellzc mentioned.

**Comment by eseliger** [comment]
SGTM, can we be sure that it’s always set for managed instances that way too? Not bound to an env var, just want to be sure we don’t forget it 😬

**Comment by michaellzc** [comment]
> SGTM, can we be sure that it’s always set for managed instances that way too? Not bound to an env var, just want to be sure we don’t forget it 😬

Yes! Just let us know which flag we need to toggle on or off depending on whether executors is enabled or not (a matrix similar to the one from https://github.com/sourcegraph/sourcegraph/issues/36920 would be great)

We invest in a lot of automation with the current MI in our `mg` cli tool.

New MI is configured with [`mg init-instance`](https://github.com/sourcegraph/deploy-sourcegraph-managed/blob/main/util/cmd/mg_init_instance.go) where we bootstrap the instance to our desire state

`mg syn-alerts` is used to make sync observability alerts

We can later build a `mg reconcile` to make changes across all configurations, e.g. site-config, feature flags, to our desired state.

**Comment by michaellzc** [comment]
>  It sounds like not having a way to detect a managed instance has been annoying for teams before

@courier-new would you share some specific examples where product teams would like to have `IsManagedInstance`?

**Comment by bobheadxi** [comment]
> cc @bobheadxi you implemented the feature flag system so figure you may want to chime in as well

Haven't looked in-depth yet but wanted to clarify that I didn't implement the feature flag system, I only added a UI for it!

**Comment by bobheadxi** [comment]
That said, my 2c:

- this new deploy type actually doesn't feel like an abuse of "deploy type" to me, at least semantically - managed instances is, indeed, a deploy type. `SourcegraphDotComMode` is not a good precedent but if somebody asked me where to put such a config I would say, "make it a deploy type"
- that said, I feel like we should not add this deploy type. What happens when we add new managed instance types, e.g. an MI 2.0? do we add another deploy type? combine flags and deploy types? what happens if Cloud becomes a priority again, and previously MI-specific feature flag checks in the code need to be shuffled around again? as mentioned, `SourcegraphDotComMode` is not a good precedent and ideally we would not have these scenarios

tl;dr I think @michaellzc's approach is the right way here: create everything under granular feature flags (Michael's example, "i.e. hide_run_batch_spec_button instead of is_managed_instance", is a good one), and `mg` can apply a set of feature flags automatically to all managed instances.

The only caveat to this approach is if, e.g. batch changes billing tracking is enabled behind a feature flag, a site admin on the customer side could (intentionally or unintentionally) disable this feature flag - we might need to introduce some capability to lock down the ability to configure (at least some) feature flags

**Comment by caugustus-sourcegraph** [comment]
I agree that using this setting to toggle functionality might not be the best approach - but it could still be a valuable change for analytics purposes. We don't seem to be able to identify managed instances in Looker; we've fallen back on looking at whether a sourcegraph email address exists, but that doesn't capture every instance.

**Comment by rafax** [comment]
+1 to making the "Enable Batch Changes" a more granular feature-flag - as @michaellzc suggested, we can find a way to set it correctly across all MI. 

Also +100 to `SourcegraphDotComMode` being a bad precedent - there's a lot of functionality that should be available outside of Cloud (but isn't), and we had at least a few cases where we needed a code change + full deploy to enable things that should have been a config change.
